### ISSUE TITLE
Optimal Pairs and other data mining

### ISSUE THREAD

**Issue by NavneethJayendran** [Togira/CS125-Research#15]
ON OPTIMAL PAIRING

We haven't really qualified what it means to be an "optimal pair". But regardless of how we define this, I see an issue with this idea. My assumption is that @pranaygp proposes trying to come up with some set S of disjoint {Student, Student} pairs that includes all students, such that S maximizes some collective metric of performance (correct me if I'm wrong though). Not only is this challenging in principle, I don't feel it is possible in practice given our data.

LIMITING CASE ANALYSIS

To illustrate this, simply consider the much easier question of, "Who is the best partner for this particular student?"

I think the current policy is that students should find different partners every lecture? That being the case, finding an optimal partner seems futile from a statistical standpoint because it would be very difficult to establish correlation. ie Even if student A performed best when paired with B on a single occasion, it's difficult to tell if that had anything to do with student B without pairing them again and seeing if the trend is consistent. And in the process of doing so, we would prevent A and B from pairing with other students and limit our data collection.

I suppose if the performance when A is paired with B were like five standard deviations above student A's mean, we could immediately identify something significant. But that seems unrealistic for the most part; we wouldn't find this happening more than a few times.

So simply because of the nature of our data collection methods, I can't see any way to discover optimal pairs between students, and certainly not with such confidence that we could make recommendations to students on how they select partners.

But if we didn't care about statistics and simply assumed that the best partner for A is B such that A performed best after getting paired with B... then perhaps we could find a set S of optimal pairs that maximizes some global metric (though defining a meaningful metric can be challenging itself, even before we consider how to maximize it). But at the end of the day, whatever result we get probably won't have much physical significance, especially since there are more than 400 students but only some 50 lectures.

ALTERNATIVE QUESTIONS

However, we can still test other general trends. For example, do students perform better after being paired with top performers (or, more generally, what correlations exist between partner performances)? Do students who always pick different partners perform better than those who tend to pick the same ones (this could be used to test whether the current partner selection procedures are a good idea)? Basically the factor that makes optimal pairing difficult is that the number of possible pairs is very large given the amount of data we receive. But quantitative relationships like this don't look at specific students and are fairly straightforward.

**Comment by pranaygp** [comment]
Your understanding of the problem is correct.

**This is a long discussion, so, in the interest of muggles and half-bloods, there is a TL:DR at the end.

* The current policy only exists so that we can get more variety in data.
* This greater variety in data only exists so that we can do certain types of data analysis with greater accuracy.
* One instance of statistical analysis that can be done better is **optimal pairing**

**Response to your first argument** :

To your first paragraph of argument, I say:  My my! The answer to life really is 42 isn't it?
Sorry, I had to make that reference somewhere :stuck_out_tongue: 

It's true that 1 pairing isn't enough to say that it's always true. Solving this is a part of solving this research question to being with. Possible solutions I can think of off the top of my head:

1. pair the same person twice. If the same pairing helps again, this is probably a good pairing.
2. pair students with other people of the same type (I have an intuitive feeling that if we draw social network type graphs of students, we will see clustering and there will be certain "types" of students) and resample data

There are more solutions you can find to this issue. There's also the case that none of them will work, in which case, the question of optimal pairing still remains the same:

> Is there such an algorithm that can use the data we have as inputs and pair students together every lecture such that the average performance of the class is as high as possible at the end of the semester? _If yes, can we find it?_

The question doesn't ask if the algorithm can be applied with the same course policies. We, using our research, declare that _there is_ an algorithm to find pairing (and maybe even find the algorithm). It's up to readers of our papers/professors to use what we find and apply it in their classes if they so see fit.

Optimization problems, and most research, in fact, isn't tied a particular use-case.

**Response to your second arguement**:

If the deviation from the mean is 5 times the standard deviation, then that's definitely crazy and they should get married and have kids and all that fun stuff.

You're not wrong about this. The lack of **consistent** data is definitely an issue that we're gonna face while trying to solve this problem, which is exactly why this the problem begins with a **IS THERE ... **

We don't know if we can find an algorithm, but using the amount of data we have, we can come close to finding a "This is probably a good pairing, but don't take my word for it" solution.

My idea for this is that, even if our algorithm is only right 10% of the time, that's still an improvement in the class average. There are a lot of factors that come into play here. It's definitely a tough problem, and we need both sides of the argument. Hopefully, we can either prove or disprove this hypothesis by the end of this research project.

**Response to your third argument**:

**THAT'S AWESOME** It's a great thing that you can't see any way to discover optimal pairings between students. If you did, this wouldn't be a hard problem AT ALL! Unfortunately, at this very moment, I can neither prove or disprove what this question proposes. Hopefully, by the end of this project, we can do either one.

Also, about the more subtle concerns you expressed within this argument, I discussed them in the previous section _Response to your second arguement_

We can definitely observe general trends too, and that's pretty exciting, but I've been super excited about the **optimal pairing** problem, and if Professor Chapman agrees, it's something I want to dedicate my time in this project to. If he does let me go forward with this, **I want you** @NavneethJayendran to help me with this research question. If you aren't interested, though, that's fine too, It's really up to you (and/or Professor Chapman) at the end of the day.

And yes, I agree. **The optimal pairing problem hinges on devising a good metric for performance**

TL;DR:

You're in a CS research team in one of the top Universities out there. You're a wizard by birth. Go read the actual arguments above and stop being lazy.

**Comment by pranaygp** [comment]
I also think we shouldn't get sidetracked from winter foundation tasks now right now. I'd love to have this discussion in person next sememster at a research meeting. Right now, I'm gonna have to defer my time to easier task and fun New Year's parties :)

**Comment by NavneethJayendran** [comment]
I definitely see your point. It would be useful to find a pairing system that optimizes class performance. But a limiting factor here is indeed the quantity of data points available through observations (feedback entries). There are around 80000 potential pairings for a class of 400 students, and only 20000 of those pairings will ever be tested (assuming 50 lectures, perfect attendance, and distinct partner selections).

At the end of the day, we can't really expect to find more than the best among those 20000. And actually acting upon these results would decrease our reservoir of information even more by restricting pairings. And this is all ignoring statistical significance, which could only be achieved by removing more and more potential data points. And if we're looking at specific individuals (as opposed to the relationships between invariant quantities like the performance of Student A versus the performance of Student B), what we find will only apply for a single semester.

But what I do agree with is that finding optimal pairs given available data (or even given arbitrarily large amounts of data) is a very hard task. The idea of finding ideal marriages in our class actually did come to mind when I was thinking about this, in the sense that I suspected that this task might be out of the scope of our abilities and resources. But if you can think of some place to start, I would definitely want to contribute. Let me know if you have any suggestions on a performance metric, because problems like this are not at all uninteresting. But it does help to assess the risks and rewards of any endeavor before we try to undertake it.

**Comment by pranaygp** [comment]
@NavneethJayendran Exactly. If you look at the way I worded the initial problem, it reflects the same concerns.

In the end, we really are only finding best (or atleast better than average) pairings among the subset of possible pairings. However, extending this to the entire set is (theoritically) not too different/difficult.

I was hoping we could all sit down and discuss ways to do this once we get back from break. I have no solid ideas on how to start answering this question yet; only random thought floating around.

**Comment by NavneethJayendran** [comment]
@pranaygp I think I understand better what you meant now. And it just occurred to me that such an algorithm could apply across semesters and work with limited data if we used general criteria to assess the quality of a pairing.

For example, if we did find a strong positive correlation between pairing males and females together and high quiz scores, we could place a positive weight on such pairings so that the algorithm can possibly make match decisions without even knowing how such a match worked out in the past. It might match Student A with Student B for the best time just based off their statistics, and we can later test if that pairing works out.

Now that I think of it, a pairing algorithm based on general information (like in an online dating site, I assume?) would be feasible. And we could add additional weights on pairs once we actually observe the results (so if A and B both performed poorly after being together, we could put a very negative weight on that pair for future runs). So I guess this procedure would favor lectures in which seating arrangements were premade? And maybe occasionally we could just give randomized seating arrangements to collect more data.

I'm actually pretty excited to see how something like that could turn out as well. Hopefully this provides some place to start. If anyone has more suggestions, we could keep this issue up for a while.

**Comment by pranaygp** [comment]
Your'e thinking along the same lines as I am.

Let's work on this once we get back. Right now, I just wanna get LectureInitializer running again so that I can port stuff over to `beaker`
